### PR TITLE
Specify the DynamicsCompressor

### PR DESCRIPTION
**Pull request by padenot**
Here is it. It should be fairly complete (I addressed the comments we made during the f2f).

### CODE REVIEW

**Comment by padenot** [comment]
Preview: https://rawgit.com/padenot/web-audio-api/4fa027223c5972b72c0d587c5eb736115d558a66/index.html#processing

**Comment by svgeesus** [comment]
This seems fairly clearly explained, to me. (A couple of typos - treashold -> threshold, Enveloppe -> Envelope).
We should probably log an actual EnvelopeFollower node as a post-v1 feature request, as it is a useful building block.

File: index.html [hunk 1]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+                </li>
+                <li>Compute <var>metering gain</var> to be <var>final
+                gain</var>, <a href="#linear-to-decibel">converted to
+                decibel</a>.
+                </li>
+              </ol>
+            </li>
+          </ol>
+          <p>
+            The makeup gain is a fixed gain stage that only depends on ratio,
```
Comment (rtoy) [inline index.html]: Maybe add `<dfn>makeup gain</dfn>` here?  And links to this from other places?

File: index.html [hunk 2]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+          </p>
+          <ol>
+            <li>Let <var>full range gain</var> be the value returned by putting
+            1.0 through the <a>compression curve</a>.
+            </li>
+            <li>Let <var>full range makeup gain</var> the inverse of <var>full
+            range gain</var>.
+            </li>
+            <li>Return the result of computing <var>full range makeup
+            gain</var><sup>0.6</sup>.
```
Comment (rtoy) [inline index.html]: This is pretty hard to read to notice that it's the 0.6 power. (I initially read it as footnote number 0.6!).  Maybe just say "take the 0.6 power of full range makeup gain"?

File: index.html [hunk 3]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+            ratio of the <var>compressor gain</var> and the <var>detector
+            average</var> through a curve. User-agents are allowed to choose
+            the shape of the attack and release curve. However, they MUST
+            respect the following constraints:
+          </p>
+          <ul>
+            <li>The input to the curve MUST be the ratio of the <var>compressor
+            gain</var> and the <var>detector average</var>.
+              <div class="note">
+                When attacking, this number inferior or equal to 1, when
```
Comment (rtoy) [inline index.html]: "inferior or equal to" -> "lis ess than or equal to"?

File: index.html [hunk 4]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+                releasing, this number is strictly superior to 1.
```
Comment (rtoy) [inline index.html]: "superior" -> "greater than"?

File: index.html [hunk 5]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+            </li>
+          </ul>
+          <div class="note">
+            It is allowed, for example, to have a compressor that performs an
+            <em>adaptative release</em>, that is, releasing faster the harder
+            the compression, or to have curves for attack and release that are
+            not of the same shape.
+          </div>
+          <p>
+            Putting a the absolute value of a sample through a <dfn>compression
```
Comment (rtoy) [inline index.html]: You use the phrase "putting foo through a curve" in many places above.  Maybe add a `<dfn>` for this and reference it?

File: index.html [hunk 6]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+            curve</dfn> means computing the absolute value of this sample to a
+            function, and returning the computed value. This function MUST
+            respects the following characteristics:
+          </p>
+          <ol>
+            <li>This function is the identify up to the value of the linear
```
Comment (rtoy) [inline index.html]: Typo: "identify" -> "identity"?

File: index.html [hunk 7]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+            <code>threshold</code> (i.e., \(f(sample) = sample\)).
```
Comment (rtoy) [inline index.html]: Maybe just use f(x) = x?  Easier to read as a mathematical formula.

File: index.html [hunk 8]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+            </li>
+            <li>From the <code>threshold</code> up to the <code>threshold +
+            knee</code>, User-Agents can choose the curve shape. The whole
+            function MUST be continuous (except as noted in the note below) and
```
Comment (rtoy) [inline index.html]: "except as noted in the note below" -> "except as noted below"?

File: index.html [hunk 9]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+            monotonically increasing.
+              <div class="note">
+                As such, if the <code>knee</code> is 0, the
+                <a>DynamicsCompressorNode</a> is a hard-knee compressor, and
+                the function has a discontinuity at <code>threshold</code>.
```
Comment (rtoy) [inline index.html]: This seems to contradict the requirement that the curve MUST be continuously differentiable.

File: index.html [hunk 10]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+              </div>
+            </li>
+            <li>This function returned is linear, based on the ratio, after the
+            <code>threshold</code> and the soft knee (i.e., \(f(sample) =
```
Comment (rtoy) [inline index.html]: Copy the formula from earlier so that it has the same form (1/ratio * x).

File: index.html [hunk 11]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+              <ul>
+                <li>The first part is the identify: \(f(x) = x\).
+                </li>
+                <li>The second part is the soft-knee portion, which MUST be a
+                monotonically increasing function.
+                </li>
+                <li>The third part is a linear function of equation \(f(x) =
+                \frac{1}{ratio} \cdot x \).
+                </li>
+              </ul>This curve MUST be continuously differentiable, and
```
Comment (padenot) [inline index.html]: I think we need to write that it is allowed to have a function that is not continuously differentiable iff the knee is 0, which kind of makes sense, because the purpose of a hard knee is to have an hard edge.

File: index.html [hunk 12]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+            </li>
+            <li>The compression curve is has three parts:
+              <ul>
+                <li>The first part is the identify: \(f(x) = x\).
+                </li>
+                <li>The second part is the soft-knee portion, which MUST be a
+                monotonically increasing function.
+                </li>
+                <li>The third part is a linear function of equation \(f(x) =
+                \frac{1}{ratio} \cdot x \).
```
Comment (padenot) [inline index.html]: I would have done that in proper LaTeX, but the layout is weird in mathjax.

File: index.html [hunk 13]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+            <a href="#associated">associated</a> with.
+            </li>
+            <li>Set <var>attack</var> and <var>release</var> to the
+            multiplication between themselves and the sample-rate for this <a>
+              BaseAudioContext</a>.
+            </li>
+            <li>For each samples <var>input</var> of <var>input signal</var>,
+            execute the following steps:
+              <ol>
+                <li>Let <var>envelope rate</var> be the result of computing the
```
Comment (padenot) [inline index.html]: `[[foo]]` refer to internal slots of objects (it's the ECMAScript notation). `<var></var>` is recommended for algorithm variables.

Comment (padenot) [inline index.html]: Also I'm switching the names around a bit so that it's less confusing.

File: index.html [hunk 14]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+                <a href="envelope-rate">envelope rate</a>
+                </li>
+                <li>Let <var>attenuation</var> be the value of putting the
```
Comment (padenot) [inline index.html]: Yes it's defined below:

> Putting a the absolute value of a sample through a <dfn>compression curve</dfn> means computing the absolute value of this sample to a function, and returning the computed value. This function MUST respects the following characteristics:

File: index.html [hunk 15]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+                absolute value of <var>input</var> <a href="compression-curve">
+                  through the compression curve</a>.
+                </li>
+                <li>Add the difference between <var>attenuation</var> and <var>
+                  detector average</var>, multiplied by the <var>enveloppe
+                  rate</var> to <var>detector average</var>, and clamp
```
Comment (padenot) [inline index.html]: I'll be spelling out this computation in mode details, so that the precedence is very clear.

File: index.html [hunk 16]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+                  <var>detector average</var> to 1.0.
+                </li>
+                <li>If attacking, add <var>detector average</var> minus
+                <var>compressor gain</var>, multiplied by <var>enveloppe
+                rate</var> to <var>compressor gain</var>, and clamp
```
Comment (padenot) [inline index.html]: Spelled it out in more detail.

File: index.html [hunk 17]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+                <a href="envelope-rate">envelope rate</a>
+                </li>
+                <li>Let <var>attenuation</var> be the value of putting the
+                absolute value of <var>input</var> <a href="compression-curve">
+                  through the compression curve</a>.
+                </li>
+                <li>Add the difference between <var>attenuation</var> and <var>
+                  detector average</var>, multiplied by the <var>enveloppe
+                  rate</var> to <var>detector average</var>, and clamp
+                  <var>detector average</var> to 1.0.
```
Comment (padenot) [inline index.html]: I made it clearer.

File: index.html [hunk 18]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+                </li>
+                <li>If attacking, add <var>detector average</var> minus
+                <var>compressor gain</var>, multiplied by <var>enveloppe
+                rate</var> to <var>compressor gain</var>, and clamp
+                <var>compressor gain</var> to 1.0.
```
Comment (padenot) [inline index.html]: Made this one clearer as well.

File: index.html (hunk 1 above)
Comment (padenot) [inline index.html]: Oops I forgot to send this one back to the main thread so that authors can read the value back and know how hard the compressor is being hit.

File: index.html (hunk 2 above)
Comment (padenot) [inline index.html]: I trust you on the how to express this in English, I learned maths in french :-)

File: index.html [hunk 19]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+            special object that behaves like an <a>AudioNode</a>, described
+            below:
+          </p>
+          <pre>
+              var delay = new DelayNode(context, {delayTime: 0.006});
+              var gain = new GainNode(context);
+              var compression = new EnveloppeFollower();
+
+              input.connect(delay).connect(gain).connect(output);
+              input.connect(compression).connect(gain.gain);</pre>
```
Comment (padenot) [inline index.html]: I added a picture as well.

File: index.html [hunk 20]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+          <p>
+            Now, the actual algorithm to determine the reduction gain, in
+            linear gain value.
+          </p>
+          <ol>
+            <li>Let <var>input signal</var> be the the block of audio to be
+            processed.
+            </li>
+            <li>Let <var>threshold</var>, <var>knee</var> have the value of the
+            <a>AudioParam</a> of the same name, <a>converted to linear unit</a>
```
Comment (padenot) [inline index.html]: Added that.

File: index.html [hunk 21]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+                <var>compressor gain</var> to 1.0.
+                </li>
+                <li>Else, multiply <var>compressor gain</var> by <var>enveloppe
+                rate</var>
+                </li>
+                <li>Compute <var>final gain</var> to be <var>compressor
+                gain</var> multiplied by the return value of <a>computing the
+                makeup gain</a>.
+                </li>
+                <li>Compute <var>metering gain</var> to be <var>final
```
Comment (padenot) [inline index.html]: I've added a thing on this, it's actually important.

File: index.html (hunk 5 above)
Comment (padenot) [inline index.html]: Fixed in a different way.

File: index.html (hunk 9 above)
Comment (padenot) [inline index.html]: I think the MUST applies to the whole sentence: it MUST (be continuously differentiable except when the knee is 0, and then that is the case there is a discontinuity at the knee).

I've changed the prose a bit.

File: index.html [hunk 22]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+            parameters).
+            </li>
+            <li>Let <var>ratio</var> have the value of the <code>ratio</code>
+            <a>Audioparam</a>, sampled at the time of processing of this block
+            (as a <a>k-rate</a> parameter).
+            </li>
+            <li>Let <var>attack</var> and <var>release</var> have the value of
+            the <a>AudioParam</a> of the same name, sampled at the time of
+            processing (those are <a>k-rate</a> parameters), converted to
+            sample-frames, based on the sample-rate of the
```
Comment (padenot) [inline index.html]: I clarified that. It's kept in float.

File: index.html [hunk 23]
```diff
@@ -9222,6 +9222,240 @@ <h3 id="example-2">
...
+            <a>BaseAudioContext</a> this <a>DynamicsCompressorNode</a> is
+            <a href="#associated">associated</a> with.
+            </li>
+            <li>Set <var>attack</var> and <var>release</var> to the
+            multiplication between themselves and the sample-rate for this <a>
+              BaseAudioContext</a>.
```
Comment (padenot) [inline index.html]: This is a mistake, this is kept in float.

**Comment by joeberkovitz** [comment]
Latest preview for review: https://rawgit.com/padenot/web-audio-api/ec6b3b2c1cbab235d57b07aa97501452254d3462/index.html

File: index.html [hunk 24]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+                <li>Let <var>attenuation</var> be the value of applying the
+                <a href="compression-curve">compression curve</a> to the
+                absolute value of <var>input</var> .
+                </li>
+                <li>Set <var>change-rate</var> to the difference between <var>
+                  attenuation</var> and <var>detector average</var>. Multiply
+                  <var>change-rate</var> by the <var>envelope rate</var>. Add
+                  <var>change-rate</var> to <var>detector average</var>, and
+                  clamp <var>detector average</var> to a maximum of 1.0.
+                </li>
```
Comment (joeberkovitz) [inline index.html]: I think there's a mistake here. In the algorithm, change-rate is not multiplied by the envelope rate, but by a different variable named `rate`, which in turn is derived from the  attenuation in dB divided by the baked-in parameter `sat_release_frames`.

File: index.html [hunk 25]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+                <li>If attacking, set <var>attack</var> to <var>detector
+                average</var> minus <var>compressor gain</var>. Set
+                <var>compressor gain</var> to the multiplication of
+                <var>attack</var> by <var>envelope rate</var>. Clamp
+                <var>compressor gain</var> to a maximum of 1.0.
+                </li>
+                <li>Else, multiply <var>compressor gain</var> by <var>envelope
```
Comment (joeberkovitz) [inline index.html]: Mistake: compressor gain is not set to the product of attack and envelope rate, it is increased by that amount. Could reword as follows:

Multiply _attack_ by _envelope rate_. Add attack to _compressor gain_.

File: index.html [hunk 26]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+                rate</var>
+                </li>
```
Comment (joeberkovitz) [inline index.html]: The clamping to 1.0 is needed in the Else clause also.

File: index.html [hunk 27]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+            </li>
+            <li>Return the result of taking 0.6 power of <var>full range makeup
+            gain</var>.
+            </li>
+          </ol>
+          <p>
+            <dfn id="#envelope-rate">Computing the envelope rate</dfn> is done
+            by applying a function to the ratio of the <var>compressor
+            gain</var> and the <var>detector average</var>. User-agents are
+            allowed to choose the shape of the attack and release curve.
```
Comment (joeberkovitz) [inline index.html]: The sudden introduction of "curves" here is confusing. I think we need a sentence or two explaining that the envelope rate is a scale factor by which some gain quantity is multiplied, in each sample frame. These successive multiplications, in turn, generate the envelope curve.

File: index.html [hunk 28]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+            <li>The attack curve MUST be a continuous, monotonically increasing
+            function in the range \([0, 1]\).
+            </li>
+            <li>The release curve MUST be a continuous, monotonically
+            decreasing function that is always greater to 1.
+            </li>
+          </ul>
+          <div class="note">
+            It is allowed, for example, to have a compressor that performs an
+            <em>adaptative release</em>, that is, releasing faster the harder
```
Comment (joeberkovitz) [inline index.html]: adaptative -> adaptive

File: index.html [hunk 29]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+              <div class="note">
+                When attacking, this number less than or equal to 1, when
+                releasing, this number is strictly greater to 1.
+              </div>
+            </li>
+            <li>The attack curve MUST be a continuous, monotonically increasing
+            function in the range \([0, 1]\).
+            </li>
+            <li>The release curve MUST be a continuous, monotonically
+            decreasing function that is always greater to 1.
```
Comment (joeberkovitz) [inline index.html]: greater than 1

File: index.html [hunk 30]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+          <ul>
+            <li>The input to the curve MUST be the ratio of the <var>compressor
+            gain</var> and the <var>detector average</var>.
+              <div class="note">
+                When attacking, this number less than or equal to 1, when
+                releasing, this number is strictly greater to 1.
+              </div>
+            </li>
+            <li>The attack curve MUST be a continuous, monotonically increasing
+            function in the range \([0, 1]\).
```
Comment (joeberkovitz) [inline index.html]: Given the above eplanation that I am suggesting, these requirements follow directly from the constraints on the value of the envelope rate.

File: index.html [hunk 31]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+          </ol>
+          <p>
+            <dfn id="#envelope-rate">Computing the envelope rate</dfn> is done
+            by applying a function to the ratio of the <var>compressor
+            gain</var> and the <var>detector average</var>. User-agents are
+            allowed to choose the shape of the attack and release curve.
+            However, they MUST respect the following constraints:
+          </p>
+          <ul>
+            <li>The input to the curve MUST be the ratio of the <var>compressor
```
Comment (joeberkovitz) [inline index.html]: Rather than the "input to the curve", this is more clearly described as the envelope rate itself (which in turn determines the curve). So... "the envelope rate MUST be calculated from the ratio of..."

File: index.html (hunk 11 above)
Comment (rtoy) [inline index.html]: If you just say continuous and piece-wise differentiable instead of continuously differentiable, everything should work out.  The hard knee is continuous and piece-wise differentiable.

File: index.html [hunk 32]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+              </ul>This curve MUST be continuously differentiable, and
+              corresponds to a target output level, based on the input level.
+            </li>
+          </ul>
+          <p>
+            Graphically, such a curve would look something like this:
+          </p>
+          <figure>
+            <img alt="Graphical representation of a compression curve" src=
+            "images/compression-curve.svg" style="width: 50%">
```
Comment (rtoy) [inline index.html]: In the diagram, you draw ratio as if it's the angle of the line after the knee.  It's not; it's the reciprocal of the slope.  Not exactly sure how to show that.

File: index.html [hunk 33]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+              var gain = new GainNode(context);
+              var compression = new EnvelopeFollower();
+
+              input.connect(delay).connect(gain).connect(output);
+              input.connect(compression).connect(gain.gain);</pre>
+          <figure>
+            <img src="images/dynamicscompressor-internal-graph.svg" alt=
+            "Schema of
+            the internal graph used by the DynamicCompressorNode">
+            <figcaption>
```
Comment (rtoy) [inline index.html]: Thanks for the diagram!  Looks really nice.  One nit:  Kind of hard to tell that the envelope follower output connects to the gain attribute of the gain node.  Not exactly sure how to best illustrate this, though.

File: index.html [hunk 34]
```diff
@@ -9144,7 +9146,8 @@ <h3 id="example-2">
               A read-only decibel value for metering purposes, representing the
               current amount of gain reduction that the compressor is applying
               to the signal. If fed no signal the value will be 0 (no gain
-              reduction).
+              reduction). When this attribute is read, return the value of the
+              private slot [[\reduction]].
```
Comment (padenot) [inline index.html]: Is it not what I've done?

File: index.html [hunk 35]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+            gain</var> and the <var>detector average</var>. User-agents are
+            allowed to choose the shape of the attack and release curve.
+            However, they MUST respect the following constraints:
+          </p>
+          <ul>
+            <li>The input to the curve MUST be the ratio of the <var>compressor
+            gain</var> and the <var>detector average</var>.
+              <div class="note">
+                When attacking, this number less than or equal to 1, when
+                releasing, this number is strictly greater to 1.
```
Comment (rtoy) [inline index.html]: "greater to" -> "greater than"

File: index.html [hunk 36]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+          <div class="note">
+            It is allowed, for example, to have a compressor that performs an
+            <em>adaptative release</em>, that is, releasing faster the harder
+            the compression, or to have curves for attack and release that are
+            not of the same shape.
+          </div>
+          <p>
+            Applying a <dfn>compression curve</dfn> to a value means computing
+            the value of this sample when passed to a function, and returning
+            the computed value. This function MUST respects the following
```
Comment (rtoy) [inline index.html]: "respects" -> "respect"

File: index.html [hunk 37]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+          </p>
+          <ol>
+            <li>This function is the identity up to the value of the linear
+            <code>threshold</code> (i.e., \(f(x) = x\)).
+            </li>
+            <li>From the <code>threshold</code> up to the <code>threshold +
+            knee</code>, User-Agents can choose the curve shape. The whole
+            function MUST be monotonically increasing and continuous except
+            when the <code>knee</code> parameter is 0, in which case a
+            discontinuity happens at the treshold.
```
Comment (rtoy) [inline index.html]: In the diagram with a hard knee, the curve is still continuous, but not differentiable; there is no discontinuity.  I think it's enough here to say that the curve in monotonic increasing and continuous.  No exception needs to be made when the knee is 0.

File: index.html [hunk 38]
```diff
@@ -9222,6 +9225,256 @@ <h3 id="example-2">
...
+            </li>
+          </ol>
+          <p>
+            Converting a value \(v\) in <dfn id="linear-to-decibel">linear gain
+            unit to decibel means</dfn> executing the following steps:
+          </p>
+          <ol>
+            <li>If <var>v</var> is equal to zero, return -1000.
+            </li>
+            <li>Else, return \( 20 \times \log_{10}{v} \)
```
Comment (rtoy) [inline index.html]: `\times` not really needed, but ok.  (Might need `\,` or something if you do remove the times, so that there's a tiny space between 20 and log.)

File: index.html (hunk 34 above)
Comment (joeberkovitz) [inline index.html]: IIRC, what you've done is to say [[reduction]] is ignored when the signal is present, and zero is returned instead.

File: index.html (hunk 27 above)
Comment (padenot) [inline index.html]: This is now well defined in the algorithm above.

File: index.html (hunk 24 above)
Comment (padenot) [inline index.html]: Yes, the idea here is that it's allowed to do a normal release, or adaptive release, etc. I'll rephrase that.

File: index.html (hunk 32 above)
Comment (padenot) [inline index.html]: Yes, this is somewhat of a language abuse. In music software, it's often represented like that, but we need to be more precise here.

File: index.html (hunk 33 above)
Comment (padenot) [inline index.html]: I tried to put a little "Gain" label, maybe we can do something better.

File: index.html [hunk 39]
```diff
@@ -9248,18 +9248,20 @@ <h3 id="example-2">
...
-            <li>The compression curve is has three parts:
+            <li>The compression curve has three parts:
               <ul>
-                <li>The first part is the identify: \(f(x) = x\).
+                <li>The first part is the identity: \(f(x) = x\).
                 </li>
                 <li>The second part is the soft-knee portion, which MUST be a
                 monotonically increasing function.
                 </li>
                 <li>The third part is a linear function of equation \(f(x) =
```
Comment (rtoy) [inline index.html]: Nit:  "linear function of equation" -> "the linear function:"

The ":" to make to match what you say in line 9253 above.

File: index.html [hunk 40]
```diff
@@ -9323,8 +9325,8 @@ <h3 id="example-2">
             </li>
           </ul>
           <p>
-            Now, the actual algorithm to determine the reduction gain, in
-            linear gain value.
+            Now, the actual algorithm to determine <var>reduction gain</var>, in
+            linear gain value, for each sample of input.
```
Comment (rtoy) [inline index.html]: This isn't a complete sentence.  Maybe just replace the final period with a colon?

File: index.html [hunk 41]
```diff
@@ -9350,28 +9352,36 @@ <h3 id="example-2">
...
               <ol>
                 <li>Let <var>envelope rate</var> be the result of <a href=
-                "envelope-rate">computing the envelope rate</a>
+                "#envelope-rate">computing the envelope rate</a>
                 </li>
-                <li>Let <var>attenuation</var> be the value of applying the
-                <a href="compression-curve">compression curve</a> to the
-                absolute value of <var>input</var> .
+                <li>If the absolute value of <var>input</var> is less than
+                0.0001, let <var>attenuation</var> be 1.0. Else, let
```
Comment (rtoy) [inline index.html]: Is this what Chrome and Firefox do today?

File: index.html [hunk 42]
```diff
@@ -9411,34 +9421,44 @@ <h3 id="example-2">
             <dfn id="#envelope-rate">Computing the envelope rate</dfn> is done
             by applying a function to the ratio of the <var>compressor
             gain</var> and the <var>detector average</var>. User-agents are
-            allowed to choose the shape of the attack and release curve.
+            allowed to choose the shape the enveloppe function.
```
Comment (rtoy) [inline index.html]: "enveloppe" -> "envelope", here and elsewhere. (Is this the French spelling?)

Comment (Mr21) [inline index.html]: Exactly :fr:

File: index.html [hunk 43]
```diff
@@ -9085,7 +9085,9 @@ <h3 id="example-2">
             <p>
               Let <var>node</var> be a new <a>DynamicsCompressorNode</a>
               object. <a href="#audionode-constructor-init">Initialize</a>
-              <var>node</var>, and return <var>node</var>.
+              <var>node</var>. Let [[\reduction]] be a private slot on this
```
Comment (joeberkovitz) [inline index.html]: It looks as though you changed the name of the slot everywhere except here.

File: index.html [hunk 44]
```diff
@@ -9222,6 +9225,274 @@ <h3 id="example-2">
...
+                </li>
+                <li>If the absolute value of <var>input</var> is less than
+                0.0001, let <var>attenuation</var> be 1.0. Else, let
+                <var>attenuation</var> be the value of applying the <a
+                href="#compression-curve">compression curve</a> to the absolute
+                value of <var>input</var>, divided by the absolute value of
+                <var>input</var>.
+                </li>
+                <li>Let <var>decibels per frame</var> be the result of dividing
+                <var>attenuationDB</var> by <var>releaseRate</var>.</li>
```
Comment (joeberkovitz) [inline index.html]: _attenuationDB_ does not appear to be defined anywhere

Comment (joeberkovitz) [inline index.html]: I think you meant to divide attenuation by the quantity known in the C++ code as `sat_release_frames`, which is the number of frames in a release time interval. Note that `releaseRate` is not computed until the following step (and it's computed from the work done in this step!)

File: index.html [hunk 45]
```diff
@@ -9222,6 +9225,274 @@ <h3 id="example-2">
...
+                <li>Set <var>relaseRate</var> be the result of
```
Comment (joeberkovitz) [inline index.html]: relaseRate -> releaseRate

File: index.html [hunk 46]
```diff
@@ -9222,6 +9225,274 @@ <h3 id="example-2">
...
+                <a href="#db-to-linear">converting <var>decibels per frames
+                </var>to linear gain</a>, and substract one.</li>
+                <li>Set <var>change-rate</var> to the difference between <var>
+                  attenuation</var> and <var>detector average</var>.
```
Comment (joeberkovitz) [inline index.html]: This where _releaseRate_ is supposed to be multiplied by the difference: the new change rate is (attenuation - detector average) * release rate.

Comment (joeberkovitz) [inline index.html]: That is, unless we're attacking.

File: index.html [hunk 47]
```diff
@@ -9222,6 +9225,274 @@ <h3 id="example-2">
...
+                </li>
+                <li>Apply the <a href="#detector-curve">detector curve</a> to
+                <var>change-rate</var></li>
+                <li>Add <var>change-rate</var> to <var>detector average</var>,
+                and clamp <var>detector average</var> to a maximum of 1.0.  </li>
+                <li>If attacking, set <var>attack</var> to <var>detector
+                average</var> minus <var>compressor gain</var>. Set
+                <var>compressor gain</var> to the multiplication of
+                <var>attack</var> by <var>envelope rate</var>. Clamp
```
Comment (joeberkovitz) [inline index.html]: The sentence "Set compressor gain to the multiplication of..." appears to be left over from before. It should be replaced by the new language in line 9382 about multiplying attack by envelope rate and adding attack to compressor gain.

File: index.html (hunk 41 above)
Comment (padenot) [inline index.html]: Yes

Comment (padenot) [inline index.html]: I think it's good enough to avoid the division by zero.

File: index.html [hunk 48]
```diff
@@ -9222,6 +9225,274 @@ <h3 id="example-2">
...
+            (as a <a>k-rate</a> parameter).
+            </li>
+            <li>Let <var>attack</var> and <var>release</var> have the value of
+            the <a>AudioParam</a> of the same name, sampled at the time of
+            processing (those are <a>k-rate</a> parameters), mutiplied by the
+            sample-rate of the <a>BaseAudioContext</a> this
+            <a>DynamicsCompressorNode</a> is <a href=
+            "#associated">associated</a> with.
+            </li>
+            <li>For each samples <var>input</var> of <var>input signal</var>,
```
Comment (rtoy) [inline index.html]: Nit: "each samples" -> "each sample"

File: index.html [hunk 49]
```diff
@@ -9222,6 +9225,274 @@ <h3 id="example-2">
...
+            execute the following steps:
+              <ol>
+                <li>Let <var>envelope rate</var> be the result of <a href=
+                "#envelope-rate">computing the envelope rate</a>
```
Comment (rtoy) [inline index.html]: Nit: End with a period, like you do elsewhere.

File: index.html [hunk 50]
```diff
@@ -9222,6 +9225,274 @@ <h3 id="example-2">
...
+                </li>
+                <li>If the absolute value of <var>input</var> is less than
+                0.0001, let <var>attenuation</var> be 1.0. Else, let
+                <var>attenuation</var> be the value of applying the <a
+                href="#compression-curve">compression curve</a> to the absolute
+                value of <var>input</var>, divided by the absolute value of
```
Comment (rtoy) [inline index.html]: This phrasing is a bit confusing because I interpreted this as dividing the absolute value of the input by the absolute value of the input.  A closer reading shows that you really mean compression curve value divided by input absolute value.  Maybe rephrase or use two sentences?

File: index.html [hunk 51]
```diff
@@ -9222,6 +9225,274 @@ <h3 id="example-2">
...
+            </li>
+            <li>Return the result of taking 0.6 power of <var>full range makeup
+            gain</var>.
+            </li>
+          </ol>
+          <p>
+            <dfn id="#envelope-rate">Computing the envelope rate</dfn> is done
+            by applying a function to the ratio of the <var>compressor
+            gain</var> and the <var>detector average</var>. User-agents are
+            allowed to choose the shape the enveloppe function.
```
Comment (rtoy) [inline index.html]: "envelope"

File: index.html [hunk 52]
```diff
@@ -9222,6 +9225,274 @@ <h3 id="example-2">
...
+            However, they MUST respect the following constraints:
+          </p>
+          <ul>
+            <li>The enveloppe rate MUST be the calculated from the ratio of the
```
Comment (rtoy) [inline index.html]: "envelope"

Also are you referring to the "envelope rate" variable?

File: index.html [hunk 53]
```diff
@@ -9222,6 +9225,274 @@ <h3 id="example-2">
...
+            </li>
+            <li>From the <code>threshold</code> up to the <code>threshold +
+            knee</code>, User-Agents can choose the curve shape. The whole
+            function MUST be monotonically increasing and continuous.
+              <div class="note">
+                If the <code>knee</code> is 0, the
+                <a>DynamicsCompressorNode</a> is called a hard-knee compressor.
+              </div>
+            </li>
+            <li>This function returned is linear, based on the ratio, after the
```
Comment (rtoy) [inline index.html]: "This function returned is linear" -> "This function is linear"

File: index.html (hunk 50 above)
Comment (padenot) [inline index.html]: I've used two sentences and an intermediate variable so make this extra-clear.

File: index.html (hunk 52 above)
Comment (padenot) [inline index.html]: I've clarified this.

File: index.html (hunk 32 above)
Comment (padenot) [inline index.html]: I've tried something else, have a look.

**Comment by padenot** [comment]
The last few commits refactor the algorithm so that things are in the right order, change the diagram, address the other comments, and tidy.
